ludwig/utils/trainer_utils.py

```diff
@@ -0,0 +1,141 @@
+        num_reductions_learning_rate: int,
+        num_increases_batch_size: int,
+        train_metrics: Dict[str, Dict[str, List[TrainerMetric]]],
+        vali_metrics: Dict[str, Dict[str, List[TrainerMetric]]],
+        test_metrics: Dict[str, Dict[str, List[TrainerMetric]]],
+        last_improvement: int,
+        last_learning_rate_reduction: int,
+        last_increase_batch_size: int,
+    ):
+        """JSON-serializable holder object that stores information related to training progress.
+
+        [train/vali/test]_metrics is a nested dictionary of TrainerMetrics: feature_name -> metric_name ->
+        List[TrainerMetrics], with one entry per training checkpoint.
+
+        Note that when a model resumes training from a checkpoint, the progress tracker is deserialized from JSON, which
+        automatically converts TrainerMetrics namedtuples into regular (epoch, steps, value) tuples.
+        """
+        self.batch_size = batch_size
+        self.epoch = epoch
+        self.steps = steps
+        self.last_improvement_steps = last_improvement_steps
+        self.last_improvement = last_improvement
+        self.last_learning_rate_reduction_steps = last_learning_rate_reduction_steps
+        self.last_learning_rate_reduction = last_learning_rate_reduction
+        self.last_increase_batch_size_steps = last_increase_batch_size_steps
+        self.last_increase_batch_size = last_increase_batch_size
+        self.learning_rate = learning_rate
+        self.best_eval_metric = best_eval_metric
+        self.best_reduce_learning_rate_eval_metric = best_reduce_learning_rate_eval_metric
+        self.last_reduce_learning_rate_eval_metric_improvement = last_reduce_learning_rate_eval_metric_improvement
+        self.best_increase_batch_size_eval_metric = best_increase_batch_size_eval_metric
+        self.last_increase_batch_size_eval_metric_improvement = last_increase_batch_size_eval_metric_improvement
+        self.num_reductions_learning_rate = num_reductions_learning_rate
+        self.num_increases_batch_size = num_increases_batch_size
+        self.train_metrics = train_metrics
+        self.vali_metrics = vali_metrics
+        self.test_metrics = test_metrics
+
+    def save(self, filepath):
+        save_json(filepath, self.__dict__)
+
+    @staticmethod
+    def load(filepath):
+        loaded = load_json(filepath)
+        return ProgressTracker(**loaded)
+
+    def log_metrics(self):
+        log_metrics = {
+            "batch_size": self.batch_size,
+            "epoch": self.epoch,
+            "steps": self.steps,
+            "last_improvement_steps": self.last_improvement_steps,
+            "learning_rate": self.learning_rate,
+            "best_valid_metric": self.best_eval_metric,
+            "num_reductions_lr": self.num_reductions_learning_rate,
+            "num_increases_bs": self.num_increases_batch_size,
+        }
+        for metrics_dict_name in [
+            "train_metrics",
+            "vali_metrics",
```

Comment:
More common to use `val_metrics` here or use the follow word`validation_metrics` ?

Reply:
Thanks for the suggestion. Changed to `validation_metrics` everywhere.